libursa/DatabaseSnapshot.cpp

```diff
@@ -229,25 +229,89 @@ void DatabaseSnapshot::execute(const Query &query,
     }
 }
 
-void DatabaseSnapshot::smart_compact(Task *task) const {
-    for (auto set : OnDiskDataset::get_taint_compatible_datasets(datasets)) {
-        std::vector<const OnDiskDataset *> candidates =
-            OnDiskDataset::get_compact_candidates(set);
+// For every dataset that we want to merge, we need to keep 8*(2**24) bytes
+// in memory (about 128MB). On the one hand, merging more datasets at once
+// is faster so we want to make this number big, OTOH on most computers RAM
+// is not unlimited so we need to make sure we won't crash the DB.
+// This should be parametrised by the DB - right now just hardcode it to 100.
+constexpr int MAX_DATASETS_TO_MERGE = 100;
 
-        if (!candidates.empty()) {
-            internal_compact(task, candidates);
-        }
-    }
+void DatabaseSnapshot::smart_compact(Task *task) const {
+    spdlog::info("Smart compact operation initiated");
+    try_compact(task, /*smart=*/true);
 }
 
 void DatabaseSnapshot::compact(Task *task) const {
-    for (auto set : OnDiskDataset::get_taint_compatible_datasets(datasets)) {
-        if (set.size() >= 2) {
-            internal_compact(task, set);
+    spdlog::info("Full compact operation initiated");
+    try_compact(task, /*smart=*/false);
+}
+
+void DatabaseSnapshot::try_compact(Task *task, bool smart) const {
+    // Try to find a best compact candidate. As a rating function, we use
+    // "number of datasets" - "average number of files", because we want to
+    // maximise number of datasets being compacted and minimise number of
+    // files being compacted (we want to compact small datasets first).
+    std::vector<const OnDiskDataset *> best_compact;
+    int64_t best_compact_value = -0xFFFFFFFFFFFF;
```

Comment:
`std::numeric_limits<int64_t>::min()`

Reply:
That's not int64_t::min tho :thinking:. OTOH maybe using int64_t min is a good idea